nni/algorithms/compression/v2/pytorch/utils/evaluator.py

```diff
@@ -0,0 +1,713 @@
+
+    def patch_optimizer_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
+        """
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+        NOTE: we only patch tasks to the first optimizer right now.
+
+        Run tasks in `before_step_tasks` before `optimizer.step()` each time.
+        Run tasks in `after_step_tasks` after `optimizer.step()` each time.
+        """
+        raise NotImplementedError
+
+    def revert_optimizer_step(self):
+        """
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+
+        Revert the optimizer step to the original one.
+        """
+        raise NotImplementedError
+
+    def register_hooks(self, hooks: List[Hook]):
+        """
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+
+        The input is a list of ``TensorHook``, ``ForwardHook``, ``BackwardHook``,
+        please view how to use ``TensorHook``, ``ForwardHook``, ``BackwardHook``.
+        This function will call ``Hook.register()`` of hook in ``hooks``, and record the hook in ``self._hooks``.
+        """
+        if not hasattr(self, '_hooks'):
+            self._hooks: List[Hook] = []
+        for hook in hooks:
+            hook.register()
+            self._hooks.append(hook)
+
+    def get_all_hooks(self) -> List[Hook]:
+        """
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+
+        Get all registered ``Hook``.
+        """
+        return getattr(self, '_hooks', [])
+
+    def remove_all_hooks(self):
+        """
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+
+        Call ``Hook.remove()`` of all ``Hook`` instances in ``self._hooks``, then clear ``self._hooks``.
+        """
+        if hasattr(self, '_hooks'):
+            for hook in self._hooks:
+                hook.remove()
+            self._hooks.clear()
+
+    def train(self, max_steps: int | None = None, max_epochs: int | None = None):
+        """
+        NOTE: NNI users can call this function to train the bound model, but not necessary.
+
+        Train the bound model with default optimization loop defined by user and only change the training duration.
+        """
+        raise NotImplementedError
+
+    def finetune(self):
+        """
+        NOTE: NNI users can call this function to finetune the bound model, but not necessary.
+
+        Finetune the bound model with default optimization loop defined by user.
+        """
+        raise NotImplementedError
+
+    def evaluate(self) -> float | Tuple[float, Any]:
+        """
+        NOTE: NNI users can call this function to evaluate the bound model, but not necessary.
+
+        Evaluate the bound model, the returned metric should be a float number,
+        or a tuple of two elements with the first element being a float number.
```

Comment:
"a tuple of two elements", what is the other element?

Reply:
The other element is usually the original metric, any type is ok.
in lightning evaluator, the second element is the returned value of `trainer.test()`, I will add more docstring to explain this.